### PR TITLE
refactor: Docset/LocProvider -> BuildOptions

### PR DESCRIPTION
**Pull request by yufeih**
This PR introduces a `BuildOptions` class to replace `Docset` and `LocProvider`.

`BuildOptions` contains information needed for build but are not derived from `Config`. It currently includes:

- Docset Path
- Repository
- Localization related properties derived from repository: locale, enable sxs, fallback docset path...

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5696)

### CODE REVIEW

**Comment by yufeih** [comment]
@mingwli these 2 failed docsets are not provisioned?

**Comment by mingwli** [comment]
> @mingwli these 2 failed docsets are not provisioned?

It's caused corresponding branches have wrong docset-name. I just updated the testing branch with right name and queued a v3 branch-build.

**Comment by mingwli** [comment]
> ⚠[docs-help-pr](https://github.com/OPS-E2E-PPE/docs-help-pr)(00:00:45.9314027 | exceed 30s, 33844 more diff)

This could be caused by concurrent server build, since I updated the config file.

File: src/docfx/lib/LocalizationUtility.cs [hunk 1]
```diff
@@ -38,89 +39,69 @@ public static string AddLeftToRightMarker(CultureInfo culture, string text)
...
                 {
-                    fallbackBranch = branchWithoutLocale;
-                }
-
-                if (TryGetContributionBranch(fallbackBranch, out var contributionBranch))
-                {
-                    fallbackBranch = contributionBranch;
+                    foreach (var branch in new[] { fallbackBranch, "master" })
+                    {
+                        if (packageResolver.TryResolvePackage(new PackagePath(fallbackRemote, branch), PackageFetchOptions.None, out var fallbackRepoPath))
```
Comment (yufeih) [inline src/docfx/lib/LocalizationUtility.cs]:
```suggestion
                        if (packageResolver.TryResolvePackage(new PackagePath(fallbackRemote, branch), Package,...., out var fallbackRepoPath))
```

Comment (yufeih) [reply src/docfx/lib/LocalizationUtility.cs]: hoops, this code is correct.

File: src/docfx/build/Build.cs [hunk 2]
```diff
@@ -122,25 +97,18 @@ private static bool BuildDocset(string docsetPath, string? outputPath, CommandLi
                 context.Output.WriteJson(".dependencymap.json", dependencyMap.ToDependencyMapModel());
                 context.Output.WriteJson(".links.json", fileLinkMap);
 
-                if (options.Legacy)
+                if (context.Config.Legacy && context.Config.OutputJson)
```
Comment (live1206) [inline src/docfx/build/Build.cs]: no need to handle non-legacy case here?

File: src/docfx/build/Build.cs [hunk 3]
```diff
@@ -41,34 +39,21 @@ public static int Run(string workingDirectory, CommandLineOptions options)
 
         private static bool BuildDocset(string docsetPath, string? outputPath, CommandLineOptions options)
         {
-            List<Error> errors;
-            Config? config = null;
+            Config? logConfig = null;
```
[the remark this answers is not in the document]
Comment (yufeih) [reply src/docfx/build/Build.cs:42]: removed

File: src/docfx/build/Build.cs (hunk 2 above)
Comment (yufeih) [reply src/docfx/build/Build.cs:100]: right

File: src/docfx/build/context/Context.cs [hunk 4]
```diff
@@ -125,6 +123,7 @@ public Context(string outputPath, ErrorLog errorLog, CommandLineOptions options,
 
         public void Dispose()
         {
+            PackageResolver.Dispose();
```
[the remark this answers is not in the document]
Comment (yufeih) [reply src/docfx/build/context/Context.cs:125]: yep

File: test/docfx.Test/build/ConfigTest.cs [hunk 5]
```diff
@@ -10,50 +10,6 @@ namespace Microsoft.Docs.Build
 {
     public static class ConfigTest
     {
-        [Theory]
```
[the remark this answers is not in the document]
Comment (yufeih) [reply test/docfx.Test/build/ConfigTest.cs:13]: most of them are testing `locale == default locale` logic, which now is replaced by a `IsLocalizedBuild` flag. Added back some of them.

**Pull request merged**